### PR TITLE
Memory-efficient attention (also code cleaned up and a colab added)

### PR DESCRIPTION
**Pull request by neonsecret**
Uses less memory, now can generate 576x1280 images with 6 gb vram

### CODE REVIEW

**Comment by DisaKzionda** [comment]
now i can run 704x704 images on my 1050ti

**Comment by basujindal** [comment]
Hi, a big thank you for the commits, I am travelling right now, will  merge these changes as soon as I get stable internet. Thanks again!

**Comment by scottmudge** [comment]
It would probably be better to remove the cleanups and "beautifying" commits from this PR. Along with the other unrelated changes (Add diffusers, colab notebooks, etc). 

They're unrelated to the attention.py change and all the whitespace and code-rearrangements are going to make future PRs and code merges a major pain.

**Comment by ZeroCool22** [comment]
We just **replace** the code on **attention.py** with the new one and that's all right?

File: LICENSE [hunk 1]
```diff
@@ -1,14 +1,82 @@
...
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
\ No newline at end of file
+Copyright (c) 2022 Robin Rombach and Patrick Esser and contributors
+
+CreativeML Open RAIL-M
```
Comment (zoru22) [inline LICENSE:3]: ?????

**Comment by zoru22** [comment]
I would not merge this PR without combing through it. There's a lot of bullshit in here. Gotta love the Rick Astley image, which makes me wonder if the attention.py changes even work.

Unfortunately I recently accidentally nuked my os install so that might be a bit :v

File: LICENSE [hunk 2]
```diff
@@ -1,14 +1,82 @@
...
+3. Grant of Patent License. Subject to the terms and conditions of this License and where and as applicable, each Contributor hereby grants to You a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable (except as stated in this paragraph) patent license to make, have made, use, offer to sell, sell, import, and otherwise transfer the Model and the Complementary Material, where such license applies only to those patent claims licensable by such Contributor that are necessarily infringed by their Contribution(s) alone or by combination of their Contribution(s) with the Model to which such Contribution(s) was submitted. If You institute patent litigation against any entity (including a cross-claim or counterclaim in a lawsuit) alleging that the Model and/or Complementary Material or a Contribution incorporated within the Model and/or Complementary Material constitutes direct or contributory patent infringement, then any patent licenses granted to You under this License for the Model and/or Work shall terminate as of the date such litigation is asserted or filed.
```
Comment (lauzadis) [inline LICENSE:40]: > sell, sell, import

typo

**Comment by qdot** [comment]
> I would not merge this PR without combing through it. There's a lot of bullshit in here. Gotta love the Rick Astley image, which makes me wonder if the attention.py changes even work.

The license update, Rick Astley image, etc are all from the original SD github. 

Core SD rickrolls you when you trigger its NSFW filter.

https://github.com/CompVis/stable-diffusion/blob/main/scripts/txt2img.py#L79

File: LICENSE (hunk 1 above)
Comment (qdot) [reply LICENSE:3]: From the CompVis SD github. This looks like it was merged then brought over here.

https://github.com/CompVis/stable-diffusion/commit/69ae4b35e0a0f6ee1af8bb9a5d0016ccb27e36dc

**Comment by satvikpendem** [comment]
Maybe it's better for @basujindal to merge the changes from the original SD repo first, and then @neonsecret can then add their changes as a PR? Because otherwise when going through the git history, it will seem as if @neonsecret changed the license and added a Rickroll image when in fact it was the original repo changing first. They say PRs should be clean from other stuff that's not specific to the main change being made.

**Comment by neonsecret** [comment]
> Maybe it's better for @basujindal to merge the changes from the original SD repo first, and then @neonsecret can then add their changes as a PR? Because otherwise when going through the git history, it will seem as if @neonsecret changed the license and added a Rickroll image when in fact it was the original repo changing first. They say PRs should be clean from other stuff that's not specific to the main change being made.

well I can roll back those changes to the essential ones and then we can merge mine and then merge the original SD

File: LICENSE (hunk 2 above)
Comment (noamnelke) [reply LICENSE:40]: > > sell, sell, import
> 
> typo

"offer to sell" and "sell" are distinct

**Comment by pastuh** [comment]
Can someone explain why everywhere I see line:
`demo.launch(share=True)`

I don't think person by default wants to share ideas. And connect to gradio.app website..

**Comment by neonsecret** [comment]
> Can someone explain why everywhere I see line: `demo.launch(share=True)`
> 
> I don't think person by default wants to share ideas. And connect to gradio website..

it's to be able access your gradio remotely 
don't worry no one will have access to it except you (or those you give the link to)

**Comment by jspraul** [comment]
47f878421c5bf97d0fff44edaa703d152cafb483 seems to be the entirety of tightening up memory usage, very nice!
```
git fetch origin pull/103/head:pull103
git cherry-pick 47f8784
```
If doesn't affect runtime or output (discussed [elsewhere](https://www.reddit.com/r/StableDiffusion/comments/x56e8x/the_optimized_stable_diffusion_repo_got_a_pr_that/)) this should probably be upstreamed.

**Comment by neonsecret** [comment]
up to you guys what changes to apply

**Comment by Doggettx** [comment]
There's another optimization you can do for less memory in there by changing

        sim = einsum('b i d, b j d -> b i j', q, k) * self.scale
        del q, k

into 

        sim = einsum('b i d, b j d -> b i j', q, k)
        sim *= self.scale
        del q, k

saves quite a bit of memory

**Comment by neonsecret** [comment]
> There's another optimization you can do for less memory in there by changing
> 
> ```
>     sim = einsum('b i d, b j d -> b i j', q, k) * self.scale
>     del q, k
> ```
> 
> into
> 
> ```
>     sim = einsum('b i d, b j d -> b i j', q, k)
>     sim *= self.scale
>     del q, k
> ```
> 
> saves quite a bit of memory

I don't think it will be noticeable
the main problem is the einsum, I don't know how to replace it because pytorch doesn't have any alternative

**Comment by Doggettx** [comment]
> I don't think it will be noticeable the main problem is the einsum, I don't know how to replace it because pytorch doesn't have any alternative

For me it allows me to go up a few steps in resolution, I can go completely insane if I do the the softmax even more split up

        sim[:4] = sim[:4].softmax(dim=-1)
        sim[4:8] = sim[4:8].softmax(dim=-1)
        sim[8:12] = sim[8:12].softmax(dim=-1)
        sim[12:] = sim[12:].softmax(dim=-1)

but not sure if it's always 16? 

I do run on a 3090 though...

**Comment by neonsecret** [comment]
for me in a bigger resolution the problem is with other modules, but l dont think rewriting the entire module is possible

**Comment by fpierfed** [comment]
FWIW I am seeing a significant performance degradation (I cherry picked 47f8784) on a 16GB M1 MacMini: from < 3s/it to >7s/it. Others might want to double check and make sure this is not just my environment.

**Comment by DisaKzionda** [comment]
> FWIW I am seeing a significant performance degradation (I cherry picked [47f8784](https://github.com/basujindal/stable-diffusion/commit/47f878421c5bf97d0fff44edaa703d152cafb483)) on a 16GB M1 MacMini: from < 3s/it to >7s/it. Others might want to double check and make sure this is not just my environment.

for me speed is the same

File: ldm/modules/attention.py [hunk 3]
```diff
@@ -174,29 +174,34 @@ def forward(self, x, context=None, mask=None):
...
             max_neg_value = -torch.finfo(sim.dtype).max
             mask = repeat(mask, 'b j -> (b h) () j', h=h)
             sim.masked_fill_(~mask, max_neg_value)
+            del mask
 
-        # attention, what we cannot get enough of
-        attn = sim.softmax(dim=-1)
+        # attention, what we cannot get enough of, by halves
+        sim[4:] = sim[4:].softmax(dim=-1)
+        sim[:4] = sim[:4].softmax(dim=-1)
```
Comment (patrickvonplaten) [inline ldm/modules/attention.py:193]: This leads to different similarity results though -> do the outputs still look good?

Comment (neonsecret) [reply ldm/modules/attention.py:193]: exactly the same

**Comment by neonsecret** [comment]
okay guys you know what, ima do a separate branch with only attention change and do a PR for it separately

**Comment by neonsecret** [comment]
https://github.com/basujindal/stable-diffusion/pull/117

**Comment by neonsecret** [comment]
closing this becuase https://github.com/basujindal/stable-diffusion/pull/117 will be more relevant

**Comment by TheEnhas** [comment]
The inpaint mask fix should have a pull request too IMO

**Comment by tusharbhutt** [comment]
Apologies... where is the Colab?

**Comment by tusharbhutt** [comment]
Oh, I tried that (I thought that the basujindal fork was a Colab now) and it fails to launch an Xterm window at the end of setting up the Colab.

**Comment by scottmudge** [comment]
> sim = einsum('b i d, b j d -> b i j', q, k)
>     sim *= self.scale
>     del q, k

I just tried this and unfortunately observed no changes to memory utilization, and a very slight decrease in performance.
